django/utils/text.py

```diff
@@ -62,6 +64,68 @@ def _generator():
     return "".join(_generator())
 
 
+class DjangoHTMLParser(HTMLParser):
+    void_elements = [  # https://html.spec.whatwg.org/#void-elements
+        "area",
+        "base",
+        "br",
+        "col",
+        "embed",
+        "hr",
+        "img",
+        "input",
+        "link",
+        "meta",
+        "source",
+        "track",
+        "wbr",
+    ]
+
+    def __init__(self, *, convert_charrefs=True):
+        super().__init__(convert_charrefs=convert_charrefs)
+        self.tags = []
+        self.output = ""
+        self.complete = False
+
+    def handle_starttag(self, tag, attrs):
+        if not self.complete:
+            self.output += self.get_starttag_text()
+            if tag not in self.void_elements:
+                self.tags.insert(0, tag)  # TODO DEQUE?
+
+    def handle_endtag(self, tag):
+        if not self.complete and tag not in self.void_elements:
+            self.output += f"</{tag}>"
+            self.tags.remove(tag)
+
+    def handle_data(self, data):
+        if self.complete:
+            return
+        if self.func == "words":
+            # TODO swap this round to find Match Groups?
```

Comment:
I was wondering if with a different regex we can combine the two lines which look something like this `output = escape(" ".join(data[: self.truncate_len]))` but have a different string as the base of the `join()`